File: lib/md2pdf/code_wbr.lua
-- Inserts <wbr> hints inside inline `code` spans at break-friendly
-- positions (path separators, snake_case underscores, dots,
-- camelCase boundaries, etc.) so narrow table cells can break a
-- long identifier between segments instead of mid-character.

local function html_escape(s)
  s = s:gsub('&', '&amp;')
  s = s:gsub('<', '&lt;')
  s = s:gsub('>', '&gt;')
  return s
end

local function inject_wbr(s)
  s = s:gsub('([/_%-.#:{}])', '%1<wbr>')
  s = s:gsub('(%l)(%u)', '%1<wbr>%2')
  return s
end

local function attr_str(attr)
  local parts = {}
  if attr.identifier and attr.identifier ~= '' then
    parts[#parts + 1] = ' id="' .. attr.identifier .. '"'
  end
  if attr.classes and #attr.classes > 0 then
    parts[#parts + 1] =
      ' class="' .. table.concat(attr.classes, ' ') .. '"'
  end
  for _, kv in ipairs(attr.attributes or {}) do
    parts[#parts + 1] =
      ' ' .. kv[1] .. '="' .. html_escape(kv[2]) .. '"'
  end
  return table.concat(parts)
end

function Code(el)
  local text = inject_wbr(html_escape(el.text))
  return pandoc.RawInline(
    'html', '<code' .. attr_str(el.attr) .. '>' .. text .. '</code>'
  )
end

File: lib/transpareo/md2pdf/filters/code_wbr.rb
# frozen_string_literal: true

require 'cgi'

module Transpareo
  module Md2pdf
    module Filters
      # Inserts <wbr> hints inside inline `code` spans at
      # break-friendly positions, so a narrow table cell can break a
      # long identifier between segments instead of overflowing.
      #
      # This rewrites the element's markup, so it has to run after
      # every filter that reads code text as plain text.
      module CodeWbr
        SEPARATORS = %r{([/_\-.#:{}])}
        CAMEL_CASE = /([a-z])([A-Z])/

        module_function

        def call(doc)
          doc.fragment.css('code').each do |code|
            next if code.parent&.name == 'pre'

            code.inner_html = inject(CGI.escapeHTML(code.text))
          end
        end

        def inject(text)
          text.gsub(SEPARATORS, '\1<wbr>').gsub(CAMEL_CASE, '\1<wbr>\2')
        end
      end
    end
  end
end
